blog/2024-08-12-ts-api-refactor.md

```diff
@@ -0,0 +1,488 @@
+
+With this refactor, the XRPC client is now more modular and reusable. Session management, retries, cryptographic signing, and other request-specific logic can be implemented in the fetch handler itself rather than by the calling code.
+
+A new abstract class named `Agent`, has been added to `@atproto/api`. This class will be the base class for all Bluesky agents classes in the `@atproto` ecosystem. It is meant to be extended by implementation that provide session management and fetch handling.
+
+As you adapt your code to these changes, make sure to use the `Agent` type wherever you expect to receive an agent, and use the `AtpAgent` type (class) only to instantiate your client. The reason for this is to be forward compatible with the OAuth agent implementation that will also extend `Agent`, and not `AtpAgent`.
+
+```ts
+import { Agent, AtpAgent } from '@atproto/api'
+
+async function setupAgent(service: string, username: string, password: string): Promise<Agent> {
+  const agent = new AtpAgent({
+    service,
+    persistSession: (evt, session) => {
+      // handle session update
+    },
+  })
+  
+  await agent.login(username, password)
+  
+  return agent
+}
+```
+
+```ts
+import { Agent } from '@atproto/api'
+
+async function doStuffWithAgent(agent: Agent, arg: string) {
+  return agent.resolveHandle(arg)
+}
+```
+
+```ts
+import { Agent, AtpAgent } from '@atproto/api'
+
+class MyClass {
+  agent: Agent
+
+  constructor () {
+    this.agent = new AtpAgent()
+  }
+}
+```
+
+## Breaking changes
+
+Most of the changes introduced in this version are backward-compatible. However, there are a couple of breaking changes you should be aware of:
+
+- Customizing `fetch`: The ability to customize the `fetch: FetchHandler` property of `@atproto/xrpc`'s `Client` and `@atproto/api`'s `AtpAgent` classes has been removed. Previously, the `fetch` property could be set to a function that would be used as the fetch handler for that instance, and was initialized to a default fetch handler. That property is still accessible in a read-only fashion through the `fetchHandler` property. That property can only be set during the instance creation. Attempting to set/get the `fetch` property will now result in an error.
+- The `fetch()` method, as well as WhatWG compliant `Request` and `Headers` constructors, must be globally available in your environment. Use a polyfill if necessary.
+- The `AtpBaseClient` has been removed. The `AtpServiceClient` has been renamed `AtpBaseClient`. Any code using either of these classes will need to be updated.
+- Instead of *wrapping* an `XrpcClient` in its `xrpc` property, the `AtpBaseClient` (formerly `AtpServiceClient`) class, created through `lex-cli`, now *extends* the `XrpcClient` class. This means that a client instance now passes the `instanceof XrpcClient` check. The `xrpc` property now returns the instance itself and has been deprecated.
```

Comment:
```suggestion
- Instead of *wrapping* an `XrpcClient` in its `xrpc` property, the `AtpBaseClient` (formerly `AtpServiceClient`) class - created through `lex-cli` - now *extends* the `XrpcClient` class. This means that a client instance now passes the `instanceof XrpcClient` check. The `xrpc` property now returns the instance itself and has been deprecated.
```